rpcs3/Emu/RSX/RSXThread.cpp

```diff
@@ -456,37 +456,57 @@ namespace rsx
 
 		last_flip_time = get_system_time() - 1000000;
 
+		vblank_count = 0;
+
 		thread_ctrl::spawn("VBlank Thread", [this]()
 		{
-			const u64 start_time = get_system_time();
+			// See sys_timer_usleep for details
+#ifdef __linux__
+			constexpr u32 host_min_quantum = 50;
+#else
+			constexpr u32 host_min_quantum = 500;
+#endif
+			u64 start_time = get_system_time();
 
-			vblank_count = 0;
+			const u64 period_time = 1000000 / g_cfg.video.vblank_rate;
+			const u64 s_p = period_time - u64{period_time >= host_min_quantum} * host_min_quantum;
 
 			// TODO: exit condition
 			while (!Emu.IsStopped() && !m_rsx_thread_exiting)
 			{
-				if (get_system_time() - start_time > vblank_count * 1000000 / 60)
+				if (get_system_time() - start_time >= period_time)
 				{
-					vblank_count++;
-					sys_rsx_context_attribute(0x55555555, 0xFED, 1, 0, 0, 0);
-					if (vblank_handler)
+					do
 					{
-						intr_thread->cmd_list
-						({
-							{ ppu_cmd::set_args, 1 }, u64{1},
-							{ ppu_cmd::lle_call, vblank_handler },
-							{ ppu_cmd::sleep, 0 }
-						});
-
-						thread_ctrl::notify(*intr_thread);
+						start_time += period_time;
+
+						sys_rsx_context_attribute(0x55555555, 0xFED, 1, 0, 0, 0);
+
+						if (isHLE)
+						{
+							vblank_count.release(vblank_count.load() + 1);
```

Comment:
Why not just post/pre-increment?

Reply:
its an atomic variable now its not the same.

Reply:
why do we not want a locked instruction though? Does that not defeat the purpose of making it atomic in the first place?

Reply:
its for memory ordering and seq cst store before launching the thread.

Reply:
And this guarentees single copy atomicity of load + release, even tho on x86 its placibo.

Reply:
Used `vblank_count.raw()++;` for now.

Reply:
Actually for pure correctness it needs se cst ordering so the value will be stored immediatly and wont get delayed.